asciimatics/renderers/charts.py

```diff
@@ -0,0 +1,486 @@
+
+    def __init__(self, height, width, functions, char="#", colour=Screen.COLOUR_GREEN,
+                 bg=Screen.COLOUR_BLACK, gradient=None, scale=None, axes=_BarChartBase.X_AXIS,
+                 intervals=None, labels=False, border=True, keys=None, gap=None):
+        """
+        :param height: The max height of the rendered image.
+        :param width: The max width of the rendered image.
+        :param functions: List of functions to chart.
+        :param char: Character to use for the bar. Defaults to '#'
+        :param colour: Colour(s) to use for the bars.  This can be a single value or list of
+            values (to cycle around for each bar). Defaults to green.
+        :param bg: Background colour to use for the bars.  This can be a single value or list of
+            values (to cycle around for each bar). Defaults to black.
+        :param gradient: Colour gradient to use for the bars.  This is a list of tuple pairs
+            specifying a threshold and a colour, or triplets to include a background colour too.
+            Defaults to no gradients.
+        :param scale: Maximum value for the bars.  This is used to scale the function values to
+            the maximum space available.  Any value over this will be truncated when drawn.
+            Defaults to the number of available characters in the chart.
+        :param axes: Which axes to draw.
+        :param intervals: Units for interval markers on the main axis. Defaults to none.
+        :param labels: Whether to draw size indication labels on the y-axis.
+        :param border: Whether to draw a border around the chart.
+        :param keys: Optional keys to name each bar on the x-axis.
+        :param gap: distance between bars. A value of None will auto-calculate (default). Minimum
+            value when auto-calculated is 1, for no gaps specify 0.
+
+        If the scale parameter is not specified, the maximum length of the bar is based on the
+        available space. A chart with no borders, no axes, no keys or labels will have a bar
+        height based solely on the width of the graph.
+
+            * Borders use 4 characters height and 6 characters width
+            * Keys use a height of 1
+            * Labels vertical bar chart use the width of the widest label plus 1 (label values
+            depend on the scale of the chart)
+            * An X_AXIS uses a height of 1
+            * A Y_AXIS uses a width of 1
+        """
+        super(VBarChart, self).__init__(height, width, functions, char, colour, bg, gradient,
+            scale, axes, intervals, labels, border, keys, gap)
+
+        self._precision = 0
+        if self._intervals:
+            point = str(self._intervals).find('.')
+            if point != -1:
+                # Interval is a float with a decimal point
+                self._precision = len(str(self._intervals)[point + 1:])
+
+    def _render_now(self):
+        int_h, int_w, start_x, start_y = self._setup_chart()
+
```

Comment:
Sorry - ran out of time to look at this today.  Something's not quite right here.  Check out the scale on the wavy bar chart when you run the bars.py demo.

Reply:
Looks the the interval logic isn't working...  Would it be easier to flip the existing BarChart logic to use vertical rather than horizontal intervals?

Reply:
I think I'm going to need a bit more feedback on this one before taking another stab. I've tried a couple of different algorithms and haven't been happy with any of them. The fact that the cursor is twice as tall as it is wide makes it always look weird.

IIRC, the examples in sample are 10 tall grids with and interval value that means 4 intervals get put to the screen. The algo I landed on made sure that the midway point was correct, but there is no way for the quarterly points to be right.

That being said, it is still entirely possible there is a bug in the algo I have.

If you can take another look at the algo and can propose something new, I'm happy to take another run at it.

If the algo is ok, I could pick a better example for bars.py so that it doesn't show off that you get rounding problems.

Reply:
There's always going to be integer maths issues on these charts unless you're really careful how you size them.

Based on the previous bug report, the key thing is that the scale is consistent with the bar drawing and that the bar overwrites the interval line when it is the interval value.  For example, the 0.5 interval marker should be exactly the same height as a bar of 0.5 value.

I think the key calculation is that your value for each line should be `value = (i + 1) * scale / int_h` rather than the stuff you do with deltas and precision rounding.

Yes, the origin lable (i.e. the "0") is in the wrong place, but I can live with that as we don't actually draw a line for a zero value.

Reply:
BTW.  It might be easier to do the loop once, by making the labels array an array of 2 tuples - one for the value and one a boolean to flag if you should draw an interval line.  That way you just draw it all from the one loop rather than iterating and calculating separately for the lines.

Reply:
I've changed it to use your suggested value calc.

I used to have a single loop, it gets messy quickly -- you can have labels without intervals and intervals without labels. You have to loop through the labels to calculate the widest label to get int_w correct. If you're ok with it, I think I'm going to leave this alone. Tests are passing, getting superstitious about not touching what's working :)